test/pkg/environment/common/monitor.go

```diff
@@ -0,0 +1,258 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/samber/lo"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+	"sigs.k8s.io/karpenter/pkg/utils/resources"
+)
+
+// Monitor is used to monitor the cluster state during a running test
+type Monitor struct {
```

Comment:
Something to consider going forward (since I definitely think that it's out-of-scope of this PR) but what do you think about ways that we could remove the monitor? Are there ways that we could use things like the test-id label to track nodes created during a given test run rather than having to track it through a "seen nodes" or counting mechanism?

Reply:
I think we'd still need something that keeps track of events, but I think that just devolves to a cluster state at some point.

Reply:
Maybe we could just add more hook-ins into the cluster state for tests only?